DESIGN.md

```diff
@@ -0,0 +1,400 @@
+type KeyType
+type RSAOptions
+type Sign1Message
+    func NewSign1Message() *Sign1Message
+    func (message *Sign1Message) MarshalCBOR() ([]byte, error)
+    func (m *Sign1Message) SigStructure(external []byte) ([]byte, error)
+    func (m *Sign1Message) Sign(rand io.Reader, external []byte, signer Signer) (err error)
+    func (message *Sign1Message) UnmarshalCBOR(data []byte) (err error)
+    func (m Sign1Message) Verify(external []byte, verifier Verifier) (err error)
+type SignMessage
+    func NewSignMessage() *SignMessage
+    func (m *SignMessage) AddSignature(s *Signature)
+    func (message *SignMessage) MarshalCBOR() ([]byte, error)
+    func (m *SignMessage) SigStructure(external []byte, signature *Signature) (ToBeSigned []byte, err error)
+    func (m *SignMessage) Sign(rand io.Reader, external []byte, signers []Signer) (err error)
+    func (message *SignMessage) UnmarshalCBOR(data []byte) (err error)
+    func (m *SignMessage) Verify(external []byte, verifiers []Verifier) (err error)
+type Signature
+    func NewSignature() (s *Signature)
+    func (s *Signature) Decode(o interface{})
+    func (s *Signature) Equal(other *Signature) bool
+type Signer
+    func NewSigner(alg *Algorithm, options interface{}) (signer *Signer, err error)
+    func NewSignerFromKey(alg *Algorithm, privateKey crypto.PrivateKey) (signer *Signer, err error)
+    func (s Signer) GetAlg() *Algorithm
+    func (s *Signer) Public() (publicKey crypto.PublicKey)
+    func (s *Signer) Sign(rand io.Reader, digest []byte) (signature []byte, err error)
+    func (s *Signer) Verifier() (verifier *Verifier)
+type Verifier
+    func (v *Verifier) Verify(digest []byte, signature []byte) (err error)
+```
+
+Here are the proposed APIs:
+
+```go
+const (
+    HeaderLabelAlgorithm         = 1
+    HeaderLabelCritical          = 2
+    HeaderLabelContentType       = 3
+    HeaderLabelKeyID             = 4
+    HeaderLabelCounterSignature  = 7
+    HeaderLabelCounterSignature0 = 9
+    HeaderLabelX5Bag             = 32
+    HeaderLabelX5Chain           = 33
+    HeaderLabelX5T               = 34
+    HeaderLabelX5U               = 35
+)
+
+const (
+    CBORTagSignMessage  = 98
+    CBORTagSign1Message = 18
+)
+
+const (
+    ContextSignature  = "Signature"
+    ContextSignature1 = "Signature1"
+)
+
+var (
+    AlgorithmPS256   *Algorithm
+    AlgorithmPS384   *Algorithm
+    AlgorithmPS512   *Algorithm
+    AlgorithmES256   *Algorithm
+    AlgorithmES384   *Algorithm
+    AlgorithmES512   *Algorithm
+    AlgorithmEd25519 *Algorithm
+)
+
+type Algorithm struct {
+    Name     string
+    Value    int
+    HashFunc crypto.Hash
+}
+
+func I2OSP(x *big.Int, n int) ([]byte, error)
+func OS2IP(x []byte) *big.Int
+
+type ProtectedHeader map[interface{}]interface{}
+    func (h *ProtectedHeader) MarshalCBOR() ([]byte, error)
+    func (h *ProtectedHeader) UnmarshalCBOR(data []byte) error
+    func (h *ProtectedHeader) SetAlgorithm(alg *Algorithm)

```

Comment:
Should there be `GetAlgorithm() (*Algorithm, error)`? This seems useful for protocols that rely on the header (together with an algorithm whitelist for example) to construct the `Verifier`.

Reply:
I've thought about this when I have `SetAlgorithm()`. If we have `GetAlgorithm()` implemented, it is too easy for developer to mis-use the `alg` to construct the verifier without proper checks. 

Please note that using the `alg` to construct a verifier is insecure since `alg` is not verified yet.

References:
- [CS-DRNT15-1 Section 3.3](https://github.com/theupdateframework/notary/blob/master/docs/resources/ncc_docker_notary_audit_2015_07_31.pdf).
- Similar vulnerability is introduced in [A cross-protocol attack on the TLS protocol](https://doi.org/10.1145/2382196.2382206) by Mavrogiannopoulos et al.
- Detailed discussions can be found at [secure-systems-lab/dsse#35](https://github.com/secure-systems-lab/dsse/issues/35).

A suggested way for faster signature verification is shown as below:
1. Get algorithm `vkAlg` from the verification key.
2. Get algorithm value from the protected header by `alg := protectedHeader[cose.HeaderLabelKeyID]`
3. Verify if `vkAlg.Value == alg`.
4. Proceed to other verification steps.

Reply:
I agree, the steps you outlined is what I had in mind. After step 3, how would you get an instance of the Algorithm for verification? Should there be some static helper function `GetAlgorithmFromLabel(label) (*Algorithm, error)`?

Reply:
The algorithm is bound with the `Verifier` and thus we can get the `*Algorithm` instance from the `Verifier`, not from the label.

In Go, a typical verification workflow looks like this as below

```go
verifier := getVerifierFromSomeWhereWithSomeKey()
msg := getSign1MessageFromSomeWhere()

// verify alg
vkAlg := verifier.Algorithm()
alg := msg.Headers.Protected[cose.HeaderLabelKeyID]
if vkAlg.Value != alg {
  return errors.New("signing algorithm mismatch")
}

// verify signature
// construct SigStructure
// construct ToBeSigned
// compute digest from ToBeSigned using the hash algorithm from `vkAlg.HashFunc`.
// verify signature by calling verifier.Verify(digest, msg.Signature) 
// ...
```

As you can observe, we don't need a `GetAlgorithmFromLabel(label)` function for the verification workflow.

Reply:
OK, then I'm happy with everything. I can't mark this thread as resolved for some reason, so just consider it resolved...